twine/auth.py

```diff
@@ -0,0 +1,80 @@
+import warnings
+import getpass
+import functools
+
+import keyring
+import munch
+
+from . import utils
+from . import exceptions
+
+
+class Resolver:
+    def __init__(self, settings, username=None, password=None):
+        self.settings = settings
+        self.input = munch.Munch(username=username, password=password)
+
+    @property
+    @functools.lru_cache()
+    def username(self):
+        return utils.get_userpass_value(
+            self.input.username,
+            self.config,
+            key='username',
+            prompt_strategy=self.username_from_keyring_or_prompt,
+        )
+
+    @property
+    @functools.lru_cache()
+    def password(self):
+        return utils.get_userpass_value(
+            self.input.password,
+            self.config,
+            key='password',
+            prompt_strategy=self.password_from_keyring_or_prompt,
+        )
+
+    @property
+    def system(self):
+        return self.settings.system
+
+    @property
+    def config(self):
+        return self.settings.repository_config
+
+    def get_username_from_keyring(self):
+        try:
+            creds = keyring.get_credential(self.system, None)
+            if creds:
+                return creds.username
```

Comment:
Small detail, but is the `if creds` necessary? If it's `None`, wouldn't that get caught and passed by `except AttributeError`?

Reply:
Yes, perhaps, but I'll avoid making that optimization as part of this change.